internal/tar_ball_composer.go

```diff
@@ -25,42 +31,91 @@ type ComposeFileInfo struct {
-		IncrementFromFiles: incrementFromFiles}
+func newComposeFileInfo(path string, fileInfo os.FileInfo, wasInBase bool,
+	header *tar.Header, updatesCount, updateRating, expectedFileSize uint64) *ComposeFileInfo {
+	return &ComposeFileInfo{path: path, fileInfo: fileInfo, wasInBase: wasInBase,
+		updateRating: updateRating, header: header, updatesCount: updatesCount, expectedSize: expectedFileSize}
+}
+
+type ComposeRatingEvaluator interface {
+	Evaluate(path string, updatesCount uint64, wasInBase bool) uint64
+}
+
+type DefaultComposeRatingEvaluator struct {
+	incrementFromFiles BackupFileList
+}
+
+func NewDefaultComposeRatingEvaluator(incrementFromFiles BackupFileList) *DefaultComposeRatingEvaluator {
+	return &DefaultComposeRatingEvaluator{incrementFromFiles: incrementFromFiles}
+}
+
+type TarFilesCollection struct {
+	files        []*ComposeFileInfo
+	expectedSize uint64
+}
+
+func newTarFilesCollection() *TarFilesCollection {
+	return &TarFilesCollection{files: make([]*ComposeFileInfo,0), expectedSize: 0}
+}
+
+func (collection *TarFilesCollection) AddFile(file *ComposeFileInfo) {
+	collection.files = append(collection.files, file)
+	collection.expectedSize += file.expectedSize
 }
 
 func (c *TarBallComposer) AddHeader(fileInfoHeader *tar.Header) {
 	c.headersToCompose = append(c.headersToCompose, fileInfoHeader)
 }
 
-func (c *TarBallComposer) AddFile(path string, fileInfo os.FileInfo, wasInBase bool, header *tar.Header, updatesCount uint64) {
-	updateRating := c.calcUpdateRating(path, updatesCount, wasInBase)
-	newFile := &ComposeFileInfo{path: path, fileInfo: fileInfo, wasInBase: wasInBase,
-		updateRating: updateRating, header: header, updatesCount: updatesCount}
+func (c *TarBallComposer) AddFile(path string, fileInfo os.FileInfo, wasInBase bool,
+	header *tar.Header, updatesCount uint64, expectedFileSize uint64) {
+	updateRating := c.composeRatingEvaluator.Evaluate(path, updatesCount, wasInBase)
+	newFile := newComposeFileInfo(path, fileInfo, wasInBase, header, updatesCount, updateRating, expectedFileSize)
 	c.files = append(c.files, newFile)
 }
 
-func (c *TarBallComposer) calcUpdateRating(path string, updatesCount uint64, wasInBase bool) uint64 {
+func (c *TarBallComposer) sortFiles() {
+	sort.Slice(c.files, func (i,j int) bool {
+		return c.files[i].updateRating < c.files[j].updateRating
+	})
+}
+
+func (c *TarBallComposer) Compose() ([]*tar.Header, []*TarFilesCollection) {
+	c.sortFiles()
+	tarFilesCollections := make([]*TarFilesCollection,0)
+	currentFilesCollection := newTarFilesCollection()
+	prevUpdateRating := uint64(0)
+
+	for _, file := range c.files {
+		// if the estimated size of the current collection exceeds the threshold,
+		// or if the updateRating just went to non-zero from zero,
+		// start packing to the new tar files collection
+		if currentFilesCollection.expectedSize > 0 &&
+			currentFilesCollection.expectedSize + file.expectedSize > c.tarSizeThreshold ||
+			prevUpdateRating == 0 && file.updateRating > 0 {
```

Comment:
It works for the first iteration. Should it work for all cases, when "prevUpdateRating == 0 && file.updateRating > 0" ?

Reply:
It was designed to divide the tars which contain files with zero update rating from the other tars. Since the files are ordered by update rating it designed to happen only once.